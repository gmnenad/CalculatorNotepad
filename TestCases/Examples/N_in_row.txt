// ** Problem: 
// What is chance to get D in row after N attempts, if chance to get one is p

// specific test values used in both simulations and formulas below
tD=5        // change to get this many in row
tN=44       // after this many draws
tp=50%      // if this is chance to get one in single draw

// ** SIMULATION in notepad - multiple lines to demonstrate what single line actually does
mVN(N,p)=vDim(N,()=>rnd≤p)  // create big vector (N=200), where each element is 1 or 0 with chance p
mOnes(v)=vSum(v)≡vLen(v)    // return true if all elements of vector are 1 , same as vSum(v)≡D
mSub(v,i,D)=v[i-D+1 … i]      // return subvector of length D, ending at position 'i' (@...,i-2,i-1,i)
mSubOnes(v,i,D)= mOnes(mSub(v,i,D))  // true if subvector of length D, ending at position 'i' has all ones
mEach(v,D)= ∑((i)=> mSubOnes(v,i,D), D-1,vLen(v)-1) // ∑ how many subvectors (in v) of len D had all ones
mOneTry(N,p,D)= mEach(mVN(N,p), D)>0 // true (1) if at least once in N draws we have D in row
mRepeat(N,p,D, nRep)=∑(()=>mOneTry(N,p,D), 1,nRep) // how many times we got D in row after nRep repeats
mSim(N,p,D, nSim) = mRepeat(N,p,D, nSim)/nSim // probability to get D in row, as percentage of successes

// ** SIMULATION in notepad - one line, equivalent to above multilines
// - mSim and mRepeat are replaced by builtin pSim() which does exactly same thing
// - mOneTry is replaced by call(), to generate big random vector once and use it many times
// - other mFuncs from above are just folded into same line, with vLen(bigVec)≡N and vLen(subVec)≡D
simND(N,p,D, nSim)= pSim(()=>call((v)=>∑((i)=>vSum(v[i-D+1…i])≡D,D-1,N-1),vDim(N,()=>rnd≤p))>0,nSim)
// ** ** SIMULATION in notepad - one line, using stack variable 'v'
simN2(N,p,D, nSim)= pSim(()=> { v=vDim(N,()=>rnd≤p); ∑((i)=>vSum(v[i-D…i-1])≡D,D,N)>0 } ,nSim)

// call notepad simulation and c# simulation with same parameters
simND(tN,tp,tD, 500)
simN2(tN,tp,tD, 500)
csSim(tN,tp,tD, 50000) 

// ** CALCULATION formula using recursion
prvNV(v,p,D)= vFor((i)=> if(i≡0,vSum(v)*(1-p), v[i-1]*p  ) , D)
prvV(p,D,k)= if (k≡0, vDim(D,(i)=>i≡0),  prvNV( prvV(p,D,k-1),p,D))
prv(p,D,k)= 1-vSum(prvV(p,D,k))

// call notepad calculation (prv) and c# calculation function (csPrv) , with same parameters
p=prv(tp,tD,tN)
csPrv(tp,tD,tN) 



//#C#
// c# simulation: prob to get at least once D in row after N attempts

    // draw N times and return true if got D times in 
    bool simOne(int N, double p, int D){
      var lastD=new bool[D];
      for(int i=0; i<N; i++){
          lastD[i%D]= nm.rnd.NextDouble()<=p;
        bool allD=true;
        for (int u=0; u<D; u++)
          if (!lastD[u]) allD=false;
        if (allD) return true;
      }
      return false;
    }

double csSim(int N, double p, int D, int nSim){
    // repeat above test many times (nSim) and find prob
    double nok = 0;
    for (int i=0; i<nSim; i++)
    if (simOne(N,p,D)) nok++;
    return nok/nSim;
}


// Three code lines result, without artificially stacking sentences behind ;
//      1) first 4 lines, including declaration, can be joined in one big line
//      2) return x[0];
//      3) }  - since ; behind return can not be omitted
double cSimND2(int N, double p, int D, int nSim, double[] nD=null, double[] x=null){
    for (x=new double[5]{0,0,0,0,p}; (null!=(nD=new double[D+1]))&&(x[3]<nSim); x[3]++)
    for(int i=0; ((x[2]=nD[D]<D?0:1)>=0)&&((x[0]+=x[2]/nSim)>=0)&&(x[2]==0)&&(i<N); i++)
        nD[D]+= (x[1]= nm.rnd.NextDouble()<=x[4]?1:0)-nD[i%D]+((nD[i%D]= x[1])>=0?0:0);
    return x[0];
}

// c# non-recursive calculation
double csPrv(double p, int D, int N){
  // create initial vector
  var v= new double[D];
  v[0]=1;
  // N times calculate new probability vector
  for (int k=0; k<N; k++){
    double sum= v.Sum();
    for (int i=D-1; i>0; i--) v[i]=v[i-1]*p;
    v[0]=sum*(1-p);
  }
  // get final result from prob vector
  double res=1;
  for (int i=0; i<D; i++) res-=v[i];
  return res;
}



#region Description of calculation
// Notepad calculation formula using recursion, three lines in notepad
// 1) prvNV(v,p,D)= vFor((i)=> if(i≡0,vSum(v)*(1-p), v[i-1]*p  ) , D)
//      find next 'v' given current 'v', where p=prob to get '1' and D= num '1s' in row we seek
//      v[i]= prob for 'i' 1s at end, so v[0]=prob xxx0, v[1]=prob xx01, v[3]=prob xx0111 etc
//      probs in 'v', vSum(v), are all given D '1s' were NOT found so far
//          - v[0]= prob of 0 at end = p(not found)*p(get 0)= vSum(v)*(1-p)
//          - v[i]= prob of i '1s' at end = p(i-1 previous '1s')*p(get 1)= v[i-1]*p
// 2) prvV(p,D,k)= if (k≡0, vDim(D,(i)=>i≡0),  prvNV( prvV(p,D,k-1),p,D))
//      find 'v' ( as described above) after k draws
//          - call recursively to find for 'k-1' draws, then call "find next" from above
//          - initial 'v' value is (1,0,0,...,0) vector of size D, for k==0
// 3) prv(p,D,k)= 1-vSum(prvV(p,D,k))
//      probability to draw at least once D '1s' in row
//      since vSum(v)=prob not to have drawn D '1s' in row in k attempts, then
//      1-vSum(v)= prob to draw at least once

// C# calculation follows same logic, only in non-recursive approach
//   - start with same (1,0,0,...,0) vector of size D
//   - N times change vector in for loop with same logic as in recursive prvNV
//   - return result same as prv() , 1- sum vector
#endregion
