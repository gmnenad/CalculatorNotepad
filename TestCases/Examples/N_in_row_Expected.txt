// ** Problem: 
// What is expected number of attempts to get D in row, if chance to get one is p

// specific test values used in both simulations and formulas below
tD=3        // change to get this many in row
tp=1/6      // if this is chance to get one in single draw


// ** SIMULATION, correct: 62 for 5 heads in a row, 258 for 3 sixes
simEX1(p,D)={
    v=vDim(D);
    c=0;
    while(vSum(v)<D){
        v[mod(c,D)]= rnd<=p;
        c=c+1;
    }
    return c;
}
simEX(p,D, nSim)= pSim(()=>simEX1(p,D),nSim) 
simEX(tp,tD,200)
csimEX(tp,tD,10000) // 62 & 258


// ** CALCULATION, still incorrect
calcEX(p,D)={ 
    pm=(2^(D-1)-1)/2^(D-1); // n=2^(D-1)=num. of 1/0, (n-1)/n= correction when not  all D
    v=vDim(D,()=>p);  // probabilities to draw, if 'all D' was NOT drawn so far. Starts with all '1s' = p
    ps=0;
    ex=0;
    for( i=D , i<1000 , i=i+1){ 
        pD= (1-ps)*vMul(v) ; // probability to have all D first time at this i-th step
        ex=ex+pD*i;  
        ps=ps+pD;
        v=v*pm;      // reduce all probabilities to draw '1' by factor pm
        v[mod(i,D)]=p;  // add initial probability 'p' for next draw 
    }
    ex
}
calcEX(tp,tD)
ccalcEX(tp,tD)

// ** FORMULA, correct (62 for 5 heads in a row, 258 for 3 sixes)
// x=p0(x+1)+p1*p0*(x+2)+p1^2*p0*(x+3)+...+p1^(N-1)*p0*(x+N) + p1^N*N
// x*(1-p0*sum(p^(i-1),1,N))= p0*sum(i*p^(i-1),1,N) + p1^N*N
// p1=p -> x = (1-p^N)/(1-p)/p^N
formEX(p,D)=(p^-D -1)/(1-p)
formEX(tp,tD)

//#C#

// probability to get D in row at least once after N attemps, given p
double csPrv(double p, int D, int N){
  // create initial vector
  var v= new double[D];
  v[0]=1;
  // N times calculate new probability vector
  for (int k=0; k<N; k++){
    double sum= v.Sum();
    for (int i=D-1; i>0; i--) v[i]=v[i-1]*p;
    v[0]=sum*(1-p);
  }
  // get final result from prob vector
  double res=1;
  for (int i=0; i<D; i++) res-=v[i];
  return res;
}

// ** SIMULATION 
int csimEX1(double p,int D){
    var v = new int[D];
    int c = 0;
    while(v.Sum()<D){
        v[c%D]= nm.rnd.NextDouble()<=p ?1:0;
        c=c+1;
    }
    return c;
}
double csimEX(double p,int D, int nSim){
    if (nSim<=0) return 0;
    double sum = 0;
    for(int i=0; i<nSim; i++){
        int nr=csimEX1(p,D);
        sum+= nr;
    }
    return sum/(double)nSim;
}

// CALC
double ccalcEX(double p, int D){
    double ex=0, ps=0;
    var v=new double[D];
    for(int u=0; u<D; u++) v[u]=p;
    for( int i=D ; i<10000 ; i=i+1){ 
        double pm=v.Aggregate(1.0,(m1,m2)=>m1*m2);
        double pD=(1-ps)*pm;
        ex = ex+ pD*i;
        ps=ps+pD;
        for(int u=0; u<D; u++)
            v[u]=(v[u]-pm)/(1-pm);
        v[i%D]=p;
    }
    return ex;
}

